src/main/kotlin/wdrollbackservice/WdRollbackServiceImpl.kt

```diff
@@ -0,0 +1,76 @@
+package wdrollbackservice
+
+import com.github.kittinunf.result.Result
+import io.reactivex.Observable
+import mu.KLogging
+import provider.NotaryPeerListProvider
+import withdrawalservice.WithdrawalTxDAO
+import java.util.stream.Collectors
+
+class WdRollbackServiceImpl(
+    private val withdrawalTransactionsDAO: WithdrawalTxDAO<String, String>,
+    private val notaryPeerListProvider: NotaryPeerListProvider
+) : WdRollbackService {
+
+    /**
+     * Handle withdrawal event to check its success
+     * @param withdrawalIrohaEventHash - iroha event hash
+     * @param ethTxStatus - eth transaction status
+     * @return event description based on the Ethereum transaction result
+     */
+    private fun processIrohaEvent(
+        withdrawalIrohaEventHash: String,
+        ethTxStatus: String?
+    ): Result<List<WdRollbackServiceOutputEvent>, Exception> {
+        return Result.of {
+            listOf(
+                WdRollbackServiceOutputEvent(
+                    withdrawalIrohaEventHash,
+                    ethFailedStatuses.contains(ethTxStatus)
+                )
+            )
+        }
+    }
+
+    /**
+     * Create and send transaction for assets refunding in Iroha in case of rollback
+     * @param wdRollbackServiceOutputEvent - Ethereum event outcome for a given Iroha withdrawal transaction
+     * @return hex representation of Iroha transaction hash or an empty string if rollback was not required
+     */
+    override fun initiateRollback(wdRollbackServiceOutputEvent: WdRollbackServiceOutputEvent): Result<String, Exception> {
+        if (!wdRollbackServiceOutputEvent.isRollbackRequired) {
+            return Result.of("")
+        }
+        return Result.of {
+            try {
+                notaryPeerListProvider.getPeerList().stream().map {
+                    val res = khttp.get("$it/ethwdrb/${wdRollbackServiceOutputEvent.irohaTxHash}")
+                    if (res.statusCode == 200) {
+                        withdrawalTransactionsDAO.remove(wdRollbackServiceOutputEvent.irohaTxHash)
+                    } else {
+                        throw Exception("Invalid response status code")
+                    }
+                    // TODO: Rewrite when implementing quorum
+                }.collect(Collectors.toList())[0].toString()
+            } catch (ex: Exception) {
+                logger.error(ex.toString())
```

Comment:
`ex.toString()` doesn't return any stack trace information. It will be impossible to know where exactly this exception happened. I'd rather use` logger.error("Some message",ex) `instead.

Reply:
Done